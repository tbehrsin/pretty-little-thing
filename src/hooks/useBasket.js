import React, { useState, createContext, useContext } from 'react';
import PropTypes from 'prop-types';

const PrivateTable = new WeakMap();

const BasketContext = createContext();

class Basket {
  constructor(items, setItems) {
    PrivateTable.set(this, {
      items,
      setItems: items => {
        PrivateTable.get(this).items = items;
        setItems(items);
      },
      copyItems: () => {
        return PrivateTable.get(this).items.map(it => ({ ...it }));
      },
    });
  }

  add(product) {
    const { copyItems, setItems } = PrivateTable.get(this);
    const items = copyItems();
    const index = items.findIndex(it => it.product === product);

    if (index === -1) {
      items.push({ product, count: 1 });
    } else {
      items[index].count++;
    }
    setItems(items);
  }

  remove(product) {
    const { copyItems, setItems } = PrivateTable.get(this);
    const items = copyItems();
    const index = items.findIndex(it => it.product === product);

    if (index !== -1) {
      items[index].count--;
      if (items[index].count === 0) {
        items.splice(index, 1);
      }
    }
    setItems(items);
  }

  getCountForProduct(product) {
    const { items } = PrivateTable.get(this);
    const item = items.find(it => it.product === product);

    if (!item) {
      return 0;
    } else {
      return item.count;
    }
  }

  setCountForProduct(product, count) {
    const { copyItems, setItems } = PrivateTable.get(this);
    const items = copyItems();

    const index = items.findIndex(it => it.product === product);

    if (index === -1) {
      if (count > 0) {
        items.push({ product, count });
      }
    } else {
      if (count === 0) {
        items.splice(index, 1);
      } else {
        items[index].count = count;
      }
    }

    setItems(items);
  }

  get subtotal() {
    const total = this.total;
    return parseFloat((total / 1.2).toFixed(2));
  }

  get vat() {
    const total = this.total;
    const subtotal = this.subtotal;
    return total - subtotal;
  }

  get total() {
    const { items } = PrivateTable.get(this);

    let total = 0;
    for (const item of items) {
      total += item.product.price * item.count;
    }
    return total;
  }

  get count() {
    const { items } = PrivateTable.get(this);

    let count = 0;
    for (const item of items) {
      count += item.count;
    }
    return count;
  }

  get items() {
    const { copyItems } = PrivateTable.get(this);
    return copyItems();
  }
}

const BasketProvider = ({ children }) => {
  const [items, setItems] = useState([]);
  const basket = new Basket(items, setItems);

  return (
    <BasketContext.Provider value={basket}>{children}</BasketContext.Provider>
  );
};

BasketProvider.propTypes = {
  children: PropTypes.node,
};

const withBasketProvider = Component => props =>
  (
    <BasketProvider>
      <Component {...props} />
    </BasketProvider>
  );

const useBasket = () => useContext(BasketContext);

export { Basket, BasketProvider, useBasket, withBasketProvider };
