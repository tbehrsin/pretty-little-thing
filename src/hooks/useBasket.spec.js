import React from 'react';
import { renderHook, act } from '@testing-library/react-hooks';
import { Basket, BasketProvider, useBasket } from './useBasket';
import products from 'fixtures/products.json';

describe('useBasket', () => {
  test('adds items', () => {
    const setItems = jest.fn();
    const items = [];

    const basket = new Basket(items, setItems);

    basket.add(products[0]);
    basket.add(products[0]);
    basket.add(products[1]);

    expect(setItems).toHaveBeenCalledTimes(3);
    expect(basket.items).toMatchSnapshot();
    expect(basket.getCountForProduct(products[0])).toEqual(2);
    expect(basket.getCountForProduct(products[1])).toEqual(1);
    expect(basket.getCountForProduct(products[3])).toEqual(0);

    basket.setCountForProduct(products[0], 15);
    basket.setCountForProduct(products[3], 5);
    expect(basket.getCountForProduct(products[0])).toEqual(15);
    expect(basket.getCountForProduct(products[1])).toEqual(1);
    expect(basket.getCountForProduct(products[3])).toEqual(5);
  });

  test('removes items', () => {
    const setItems = jest.fn();
    const items = [];

    const basket = new Basket(items, setItems);

    basket.add(products[0]);
    basket.add(products[0]);
    basket.add(products[1]);
    basket.remove(products[0]);

    expect(setItems).toHaveBeenCalledTimes(4);
    expect(basket.items).toMatchSnapshot();

    basket.remove(products[0]);
    expect(setItems).toHaveBeenCalledTimes(5);
    expect(basket.items).toMatchSnapshot();
  });

  test('copies items array', () => {
    const setItems = jest.fn();
    const items = [];

    const basket = new Basket(items, setItems);

    basket.add(products[0]);
    expect(items).toMatchSnapshot();

    basket.remove(products[0]);
    expect(items).toMatchSnapshot();

    expect(items).not.toBe(basket.items);
  });

  test('basket gets recreated on each render', () => {
    const wrapper = BasketProvider;
    const { result } = renderHook(() => useBasket(), { wrapper });

    const basket = result.current;
    act(() => result.current.add(products[0]));

    expect(basket).not.toBe(result.current);
    expect(basket).toBeInstanceOf(Basket);
    expect(result.current).toBeInstanceOf(Basket);

    act(() => result.current.add(products[0]));

    expect(basket.items).toEqual([{ count: 1, product: products[0] }]);
    expect(result.current.items).toEqual([
      {
        count: 2,
        product: products[0],
      },
    ]);
  });

  test('totals are correct', () => {
    const setItems = jest.fn();
    const items = [];

    const basket = new Basket(items, setItems);

    for (const product of products) {
      basket.add(product);
    }

    expect(basket.count).toMatchSnapshot();
    expect(basket.subtotal).toMatchSnapshot();
    expect(basket.vat).toMatchSnapshot();
    expect(basket.total).toMatchSnapshot();
  });
});
