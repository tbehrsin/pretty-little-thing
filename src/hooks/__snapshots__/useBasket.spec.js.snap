// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useBasket adds items 1`] = `
Array [
  Object {
    "count": 2,
    "product": Object {
      "colour": "Black",
      "id": 1,
      "img": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
      "name": "Black Sheet Strappy Textured Glitter Bodycon Dress",
      "price": 10,
    },
  },
  Object {
    "count": 1,
    "product": Object {
      "colour": "Stone",
      "id": 2,
      "img": "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024",
      "name": "Stone Ribbed Strappy Cut Out Detail Bodycon Dress",
      "price": 4,
    },
  },
]
`;

exports[`useBasket copies items array 1`] = `Array []`;

exports[`useBasket copies items array 2`] = `Array []`;

exports[`useBasket removes items 1`] = `
Array [
  Object {
    "count": 1,
    "product": Object {
      "colour": "Black",
      "id": 1,
      "img": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
      "name": "Black Sheet Strappy Textured Glitter Bodycon Dress",
      "price": 10,
    },
  },
  Object {
    "count": 1,
    "product": Object {
      "colour": "Stone",
      "id": 2,
      "img": "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024",
      "name": "Stone Ribbed Strappy Cut Out Detail Bodycon Dress",
      "price": 4,
    },
  },
]
`;

exports[`useBasket removes items 2`] = `
Array [
  Object {
    "count": 1,
    "product": Object {
      "colour": "Stone",
      "id": 2,
      "img": "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024",
      "name": "Stone Ribbed Strappy Cut Out Detail Bodycon Dress",
      "price": 4,
    },
  },
]
`;

exports[`useBasket totals are correct 1`] = `4`;

exports[`useBasket totals are correct 2`] = `32.49`;

exports[`useBasket totals are correct 3`] = `6.5`;

exports[`useBasket totals are correct 4`] = `38.99`;
